sql/src/main/java/org/apache/druid/sql/calcite/rule/FilterDecomposeConcatRule.java

```diff
@@ -0,0 +1,295 @@
+    }
+  }
+
+  /**
+   * Generate an expression for the case where matching is impossible.
+   *
+   * This expression might be FALSE and might be UNKNOWN depending on whether any of the inputs are null. Use the
+   * construct "x IS NULL AND UNKNOWN" for each arg x to CONCAT, which is FALSE if x is not null and UNKNOWN is x
+   * is null. Then OR them all together, so the entire expression is FALSE if all args are not null, and UNKNOWN if any arg is null.
+   *
+   * @param nonLiterals non-literal arguments to CONCAT
+   */
+  private static RexNode impossibleMatch(final List<RexNode> nonLiterals, final RexBuilder rexBuilder)
+  {
+    if (NullHandling.sqlCompatible()) {
+      // This expression might be FALSE and might be UNKNOWN depending on whether any of the inputs are null. Use the
+      // construct "x IS NULL AND UNKNOWN" for each arg x to CONCAT, which is FALSE if x is not null and UNKNOWN if
+      // x is null. Then OR them all together, so the entire expression is FALSE if all args are not null, and
+      // UNKNOWN if any arg is null.
+      final RexLiteral unknown =
+          rexBuilder.makeNullLiteral(rexBuilder.getTypeFactory().createSqlType(SqlTypeName.BOOLEAN));
+      return RexUtil.composeDisjunction(
+          rexBuilder,
+          Iterables.transform(
+              nonLiterals,
+              operand -> rexBuilder.makeCall(
+                  SqlStdOperatorTable.AND,
+                  rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL, operand),
+                  unknown
+              )
+          )
+      );
+    } else {
+      return rexBuilder.makeLiteral(false);
+    }
+  }
+
+  /**
+   * Given a literal (which may be wrapped in a cast), remove the cast call (if any) and read it as a string.
+   * Returns null if the rex can't be read as a string.
+   */
+  @Nullable
+  private static String getAsString(final RexNode rexNode)
+  {
+    if (!SqlTypeFamily.STRING.contains(rexNode.getType())) {
+      // We don't expect this to happen, since this method is used when reading from RexNodes that are expected
+      // to be strings. But if it does (CONCAT operator that accepts non-strings?), return null so we skip the
+      // optimization.
+      return null;
+    }
+
+    // Get matchValue from the matchLiteral (remove cast call if any, then read as string).
+    final RexNode matchLiteral = RexUtil.removeCast(rexNode);
+    if (SqlTypeFamily.STRING.contains(matchLiteral.getType())) {
+      return RexLiteral.stringValue(matchLiteral);
+    } else if (SqlTypeFamily.NUMERIC.contains(matchLiteral.getType())) {
+      return String.valueOf(RexLiteral.value(matchLiteral));
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Count the number of occurrences of substring in string.
+   */
+  private static int countOccurrences(final String string, final String substring)
+  {
+    int count = 0;
+    int i = -1;
+
+    while ((i = string.indexOf(substring, i + 1)) >= 0) {
```

Comment:
nit: should this add substring length (after matching) to i instead of just 1 inside of the match to skip further ahead in the string?

Reply:
deep question!

I thought about this for a bit and realized the current code is more correct, because of the case `CONCAT(x, '--', y) = '2---3'`. It must be recognized as ambiguous, so `countOccurrences("2---3", "--")` should return 2, not 1. I added a test case for this.

While adding more tests I also found and fixed a bug where an empty string in the list, like `CONCAT(x, '', y)` would cause an infinite loop.